[Tablet Orders] Open order search results in split view

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12170
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Previously we opened order search results modally.

With the new side-by-side view for order editing, it looked strange to open a new full screen modal over the formSheet search modal when shown on an iPad, when the merchant tapped edit from a search result.

This commit changes the search results to open in the split view context, when the side-by-side view is enabled.

Since they’re opening in the split view context, opening the side-by-side order editing screen above makes sense.

Note that this change does mean that after selecting an order, the search context is lost. Previously it could be accessed by tapping back. Ideally, we'd show the whole search in the split view, but this is an initial fix to get us ready for release. Future changes covered in #12202

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Enable the `sideBySideViewForOrderForm` feature flag
2. Launch the app on an iPad or large iPhone in landscape
3. Navigate to the `Orders` tab
4. Tap the magnifying glass
5. Search for an old order
6. Select it in the search result list
7. Observe that it's opened in the split view controller.

Repeat on a portrait or small iPhone

The selected order should be shown in both the list and the details view, but sometimes the list data source is not updated in time. In that case, 7050aa3 ensures that we don't leave a confusing selection state in the list.

Note that some of the updated functions are used for deeplinks, so some deeplink testing would also be worthwhile, however the row selection approach was already not great.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://github.com/woocommerce/woocommerce-ios/assets/2472348/b508c27c-a04a-4f0e-a820-3243aa0c6fdc


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald committed:
12170 Open order search results in split view

Previously we opened order search results modally.

With the new side-by-side view for order editing, it looked strange to open a new full screen modal over the formSheet search modal when shown on an iPad, when the merchant tapped edit from a search result.

This commit changes the search results to open in the split view context, when the side-by-side view is enabled.

Since they’re opening in the split view context, opening the side-by-side order editing screen above makes sense.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
12170 Deselect on selection failure

Older orders are often not present in the datasource, when we try to select them.

If we have another row selected, and fail to select the new one, it is often confusing to keep that unrelated order selected, as we’re still showing the new order.

This change deselects the current selection in that situation, so at least we aren’t adding confusion for merchants.

joshheald committed:
12170 Fix overscrolling issue

Previously, it could end up repeatedly scrolling and triggering fetches right back to the time you got the order. This change makes it scroll once only.

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/OrdersRootViewController.swift:
```diff
@@ -124,15 +124,34 @@ final class OrdersRootViewController: UIViewController {
+        let detailsViewController = OrderDetailsViewController(viewModel: viewModel)
+
+        viewController.navigationController?.pushViewController(detailsViewController, animated: true)
+        ServiceLocator.analytics.track(event: WooAnalyticsEvent.Orders.orderOpen(
```
We can remove the ServiceLocator call, since analytics is already declared in the top level:

```suggestion
        analytics.track(event: WooAnalyticsEvent.Orders.orderOpen(
```

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Search/Order/OrderSearchUICommand.swift:
```diff
@@ -76,13 +80,7 @@ final class OrderSearchUICommand: SearchUICommand {
-        ServiceLocator.analytics.track(event: WooAnalyticsEvent.Orders.orderOpen(
-            order: model,
-            horizontalSizeClass: UITraitCollection.current.horizontalSizeClass))
+        onSelectSearchResult(model, viewController)
```
Great change moving the navigation responsibility to the root view controller 💯

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift:
```diff
@@ -532,14 +533,18 @@ private extension OrderListViewController {
-    func highlightSelectedRowIfNeeded() {
+    func highlightSelectedRowIfNeeded(shouldScrollIfNeeded: Bool = false) {
         guard let selectedOrderID, let orderIndexPath = indexPath(for: selectedOrderID) else {
+            tableView.deselectSelectedRowWithAnimation(true)
```
This is a nice detail for error handling. Also TIL about `deselectSelectedRowWithAnimation`

joshheald replied:
Yeah, it's our extension, not something Apple provide but it hides away unwrapping an index path so it's good for this situation.

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift:
```diff
@@ -615,13 +620,14 @@ extension OrderListViewController {
-                onOrderSelected(id: order.orderID)
+                onOrderSelected(id: order.orderID, shouldScrollIfNeeded: shouldScrollIfNeeded)
             }
+            onCompletion?(hasBeenSelected)
```
🤔 Interesting, so we were not calling the completion handler when `hasBeenSelected` returns false? 

Unrelated to this PR, but around the same logic: I find the current API for `switchDetailsHandler` very confusing/opaque, returning objects like `switchDetailsHandler([], 0, true, nil)` says nothing to the consumer of the API and perhaps we should improve the interface/signatures. I'll add to my TODO list to investigate a bit around this and open an issue.

Along with this, I see we're not currently testing some of the classes involved, opened a new issue here for reference: https://github.com/woocommerce/woocommerce-ios/issues/12204

joshheald replied:
> 🤔 Interesting, so we were not calling the completion handler when hasBeenSelected returns false?

Nope, we were _never_ calling it, not just when it was false.

> Unrelated to this PR, but around the same logic: I find the current API for switchDetailsHandler very confusing/opaque, returning objects like switchDetailsHandler([], 0, true, nil) says nothing to the consumer of the API and perhaps we should improve the interface/signatures. I'll add to my TODO list to investigate a bit around this and open an issue.

Agreed. It's from the original HACK week project so it would be good to improve it at some point, because it's not clear. A lot of the issues come from it being a closure, and not being able to name the parameters, but even if we simply wrapped some of the calls in friendlier function names, it would be an improvement.

iamgabrielma reviewed (approved):
Works great! 🚢

joshheald committed:
Update WooCommerce/Classes/ViewRelated/Orders/OrdersRootViewController.swift

Co-authored-by: Gabriel Maldonado <gabriel.maldonado@automattic.com>
